src/com/ms/silverking/cloud/dht/management/OpCompletionTracker.java

```diff
@@ -17,111 +17,111 @@
+        if (!complete) {
+          incompleteOps.add(uuid);
         }
-        if (opCompletionState == null) {
-            // This completion state was removed immediately after we added it. Ignore.            
-        } else {
-            opCompletionState.setComplete(result);
+        if (opCompletionState.getResult().hasFailed()) {
+          failedOps.add(uuid);
         }
+      }
+    } finally {
+      for (UUIDBase uuid : uuids) {
+        completionState.remove(uuid);
+      }
     }
-    
-    private static class OpCompletionState {
-        private OpResult    result;
-        private Lock        lock;
-        private Condition    cv;
-        
-        OpCompletionState() {
-            this.lock = new ReentrantLock();
-            this.cv = lock.newCondition();
-            result = OpResult.INCOMPLETE;
-        }
-        
-        boolean waitForCompletion(long time, TimeUnit unit) {
-            Timer    timer;
-            
-            timer = new SimpleTimer(unit, time);
-            lock.lock();
-            try {
-                while (!result.isComplete() && !timer.hasExpired()) {
-                    try {
-                        timer.await(cv);
-                    } catch (InterruptedException e) {
-                    }
-                }
-            } finally {
-                lock.unlock();
-            }
-            return result.isComplete();
-        }
-        
-        void setComplete(OpResult result) {
-            lock.lock();
-            try {
-                if (!this.result.isComplete()) {
-                    this.result = result;
-                }
-                cv.signalAll();
-            } finally {
-                lock.unlock();
-            }
+    return new Pair<>(incompleteOps, failedOps);
+  }
+
+  public void setComplete(UUIDBase uuid, OpResult result) {
+    OpCompletionState opCompletionState;
+
+    opCompletionState = completionState.get(uuid);
+    if (opCompletionState == null) {
+      // We allow for the slight chance of a leak for now
+      completionState.putIfAbsent(uuid, new OpCompletionState());
+      opCompletionState = completionState.get(uuid);
+    }
+    if (opCompletionState == null) {
+      // This completion state was removed immediately after we added it. Ignore.
+    } else {
+      opCompletionState.setComplete(result);
+    }
+  }
+
+  private static class OpCompletionState {
+    private OpResult result;
+    private Lock lock;
+    private Condition cv;
+
+    OpCompletionState() {
+      this.lock = new ReentrantLock();
+      this.cv = lock.newCondition();
+      result = OpResult.INCOMPLETE;
+    }
+
+    boolean waitForCompletion(long time, TimeUnit unit) {
+      Timer timer;
+
+      timer = new SimpleTimer(unit, time);
+      lock.lock();
+      try {
+        while (!result.isComplete() && !timer.hasExpired()) {
+          try {
+            timer.await(cv);
+          } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)